cli/cli_app.go

```diff
@@ -542,26 +542,44 @@ func runTerragruntWithConfig(terragruntOptions *options.TerragruntOptions, terra
 		return err
 	}
 
-	return runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() error {
+	err, output := runActionWithHooks("terraform", terragruntOptions, terragruntConfig, func() (error, string) {
 		return runTerraformWithRetry(terragruntOptions)
 	})
+	if err != nil && isErrorRequiringInit(output, err, terragruntOptions) {
+		initOptions := terragruntOptions.Clone(terragruntOptions.TerragruntConfigPath)
+		initOptions.TerraformCommand = "init"
+		initOptions.TerraformCliArgs = []string{"init"}
+		initOptions.AutoInit = false
+		if err := runTerragruntWithConfig(initOptions, terragruntConfig, allowSourceDownload); err != nil {
```

Comment:
Can this get into an infinite loop? Since `runTerragruntWithConfig` is calling itself, if `init` fails again and again due to some permanent issue, won't this keep running forever?

Reply:
Yea, thinking about this more, I think this needs to use the max retry and timeout logic...

Reply:
This is a very good point. Although, in this case we have an exit condition here:

https://github.com/gruntwork-io/terragrunt/blob/94b962bb85b58bee0019c0718c0987bd92179cda/cli/cli_app.go#L526-L534

The first moment we got an error when running `terraform init`, then we would exit. This is because if the command is `init` and the `prepareInitCommand()` runs and returns an error, then this error would bubble up the stack and then we would exit.

Interestingly, I need to re-factor this, as we _might_ want to retry `terraform init` itself, which we don't at the moment.

Reply:
I am going to retire this idea of running `runTerragruntWithConfig()` recursively. It's unnecessarily complex and hard to test.